src/conditions/is_log.rs

```diff
@@ -63,14 +63,14 @@ mod test {
 
         assert_eq!(cond.check(&Event::from("just a log")), true);
         assert_eq!(
-            cond.check(&Event::from(Metric {
-                name: "test metric".to_string(),
-                namespace: None,
-                timestamp: None,
-                tags: None,
-                kind: MetricKind::Incremental,
-                value: MetricValue::Counter { value: 1.0 },
-            })),
+            cond.check(&Event::from(Metric::new(
+                "test metric".to_string(),
+                None,
+                None,
+                None,
+                MetricKind::Incremental,
+                MetricValue::Counter { value: 1.0 },
+            ))),
```

Comment:
My only real concern here is that this is a lot of args for a constructor (i.e. `None, None, None`), which can make it more difficult to understand at a glance. It's not a huge deal, but if you see any common patterns it could be worth adding a couple of more focused constructors. The other option would be a builder of some kind, but I'm not sure it's worth that yet.

Reply:
I know, that constructor is ugly. I put it in as a convenience for converting code instead of writing an even uglier multi-level struct instantiation all over the place:
```rust
Metric {
    series: MetricSeries {
        name: MetricName {
            name: "test metric".to_string(),
            namespace: None,
        },
        tags: None,
    },
    data: MetricData {
        timestamp: None,
        kind: MetricKind::Incremental,
        value: MetricValue::Counter { value: 1.0 },
    },
}
```
I'm open to ideas on how to make it better. Note that doing so means rewriting over 300 call sites. :cry:

Reply:
This seems like a good time to plug https://comby.dev/ or IntelliJ's structural search/replace :smile: 

I definitely think that we should keep using some kind of constructor or builder just to keep us decoupled from the internal representation. Since you always need a name, kind, and value, one option would be to have a constructor for those and then builder-style `with_{namespace,timestamp,tags}` methods that consume and return the metric and can be chained. Another option would be to see how many combinations there are and just add a couple of additional constructors to cover the majority of them.

If you start with the most specific pattern you're replacing (i.e. three `None`s, then two, then one), you _should_ be able to do a transformation like that in a pretty automated way.

Reply:
Looks like a good fit the dervive_builder crate

Reply:
I reduced the `fn new` parameters down to the minimum non-optional values, and added builder-like functions for the other three, and it dropped almost 400 lines of code after formatting. That's quite the win. Thanks for pointing me to comby!